### PR TITLE
Exclude 'vendor' folder with new property sonar.go.exclusions

### LINKED ISSUE #141: Exclude "vendor" folder
Since Go 1.6, go projects can contain some dependencies in a `vendor` folder.
SonarGo analysis should ignore this folder, out of the box.
Like SonarJS did for `node_modules` folder:
[JavaScriptPlugin.java#L64](https://github.com/SonarSource/SonarJS/blob/bfe406f8cf201689bd180449c34d76889d0c5dd2/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptPlugin.java#L64)
[JavaScriptExclusionsFileFilter.java](https://github.com/SonarSource/SonarJS/blob/master/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptExclusionsFileFilter.java)

Current workaround:
```
-Dsonar.exclusions="**/vendor/**" -Dsonar.test.exclusions="**/vendor/**"
```

### PR DESCRIPTION
**Pull request by vilchik-elena**
fixes #141

### CODE REVIEW

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  @Test
+  public void should_exclude_using_custom_path_regex() throws Exception {
+    MapSettings settings = new MapSettings();
+
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, "**/lib/**");
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+
+    assertEquals(true, filter.accept(inputFile("file.go")));
+    assertEquals(true, filter.accept(inputFile("vendor/file.go")));
+    assertEquals(false, filter.accept(inputFile("lib/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGQ&open=AWMWLXE6rUQoN0o0cVGQ)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGQ

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+class GoExclusionsFileFilterTest {
+
+  @Test
+  public void should_exclude_vendor_dir() throws Exception {
+    MapSettings settings = new MapSettings();
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, GoPlugin.EXCLUSIONS_DEFAULT_VALUE);
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+    assertEquals(true, filter.accept(inputFile("file.go")));
+    assertEquals(false, filter.accept(inputFile("vendor/file.go")));
+    assertEquals(false, filter.accept(inputFile("vendor/someDir/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGG&open=AWMWLXE6rUQoN0o0cVGG)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGG

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+  @Test
+  public void should_exclude_using_custom_path_regex() throws Exception {
+    MapSettings settings = new MapSettings();
+
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, "**/lib/**");
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+
+    assertEquals(true, filter.accept(inputFile("file.go")));
+    assertEquals(true, filter.accept(inputFile("vendor/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGP&open=AWMWLXE6rUQoN0o0cVGP)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGP

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    assertEquals(false, filter.accept(inputFile("someDir/lib/file.go")));
+  }
+
+  @Test
+  public void should_ignore_empty_path_regex() throws Exception {
+    MapSettings settings = new MapSettings();
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, "," + GoPlugin.EXCLUSIONS_DEFAULT_VALUE + ",");
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+
+    assertEquals(true, filter.accept(inputFile("file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGS&open=AWMWLXE6rUQoN0o0cVGS)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGS

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    assertEquals(false, filter.accept(inputFile("vendor/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGT&open=AWMWLXE6rUQoN0o0cVGT)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGT

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+  @Test
+  public void should_exclude_vendor_dir() throws Exception {
+    MapSettings settings = new MapSettings();
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, GoPlugin.EXCLUSIONS_DEFAULT_VALUE);
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+    assertEquals(true, filter.accept(inputFile("file.go")));
+    assertEquals(false, filter.accept(inputFile("vendor/file.go")));
+    assertEquals(false, filter.accept(inputFile("vendor/someDir/file.go")));
+    assertEquals(false, filter.accept(inputFile("someDir/vendor/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGH&open=AWMWLXE6rUQoN0o0cVGH)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGH

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  public void should_exclude_using_custom_path_regex() throws Exception {
+    MapSettings settings = new MapSettings();
+
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, "**/lib/**");
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+
+    assertEquals(true, filter.accept(inputFile("file.go")));
+    assertEquals(true, filter.accept(inputFile("vendor/file.go")));
+    assertEquals(false, filter.accept(inputFile("lib/file.go")));
+    assertEquals(false, filter.accept(inputFile("someDir/lib/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGR&open=AWMWLXE6rUQoN0o0cVGR)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGR

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  @Test
+  public void should_exclude_using_custom_path_regex() throws Exception {
+    MapSettings settings = new MapSettings();
+
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, "**/lib/**");
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+
+    assertEquals(true, filter.accept(inputFile("file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGO&open=AWMWLXE6rUQoN0o0cVGO)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGO

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    assertEquals(false, filter.accept(inputFile("vendor/someDir/file.go")));
+    assertEquals(false, filter.accept(inputFile("someDir/vendor/file.go")));
+  }
+
+  @Test
+  public void should_exclude_only_go() throws Exception {
+    MapSettings settings = new MapSettings();
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, GoPlugin.EXCLUSIONS_DEFAULT_VALUE);
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+    assertEquals(false, filter.accept(inputFile("vendor/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGI&open=AWMWLXE6rUQoN0o0cVGI)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGI

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  @Test
+  public void should_include_vendor_when_property_is_overridden() throws Exception {
+    MapSettings settings = new MapSettings();
+
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, "");
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+
+    assertEquals(true, filter.accept(inputFile("file.go")));
+    assertEquals(true, filter.accept(inputFile("vendor/file.go")));
+    assertEquals(true, filter.accept(inputFile("vendor/someDir/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGM&open=AWMWLXE6rUQoN0o0cVGM)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGM

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    assertEquals(true, filter.accept(inputFile("someDir/vendor/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGN&open=AWMWLXE6rUQoN0o0cVGN)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGN

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  @Test
+  public void should_include_vendor_when_property_is_overridden() throws Exception {
+    MapSettings settings = new MapSettings();
+
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, "");
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+
+    assertEquals(true, filter.accept(inputFile("file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGK&open=AWMWLXE6rUQoN0o0cVGK)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGK

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    assertEquals(true, filter.accept(inputFile("vendor/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGL&open=AWMWLXE6rUQoN0o0cVGL)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGL

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+import static org.junit.Assert.assertEquals;
+
+class GoExclusionsFileFilterTest {
+
+  @Test
+  public void should_exclude_vendor_dir() throws Exception {
+    MapSettings settings = new MapSettings();
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, GoPlugin.EXCLUSIONS_DEFAULT_VALUE);
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+    assertEquals(true, filter.accept(inputFile("file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGE&open=AWMWLXE6rUQoN0o0cVGE)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGE

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    assertEquals(false, filter.accept(inputFile("vendor/file.go")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGF&open=AWMWLXE6rUQoN0o0cVGF)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGF

File: sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    assertEquals(false, filter.accept(inputFile("someDir/vendor/file.go")));
+  }
+
+  @Test
+  public void should_exclude_only_go() throws Exception {
+    MapSettings settings = new MapSettings();
+    settings.setProperty(GoPlugin.EXCLUSIONS_KEY, GoPlugin.EXCLUSIONS_DEFAULT_VALUE);
+    GoExclusionsFileFilter filter = new GoExclusionsFileFilter(settings.asConfig());
+    assertEquals(false, filter.accept(inputFile("vendor/file.go")));
+    assertEquals(true, filter.accept(inputFile("vendor/file.json")));
```
Comment (SonarTech) [inline sonar-go-plugin/src/test/java/org/sonar/go/plugin/GoExclusionsFileFilterTest.java]: Code Smell Code Smell: Remove or correct this assertion. ([squid:S2701](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2701))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.go%3Asonar-go&pullRequest=292&issues=AWMWLXE6rUQoN0o0cVGJ&open=AWMWLXE6rUQoN0o0cVGJ)

[ISSUE_KEY]: AWMWLXE6rUQoN0o0cVGJ

**Review by saberduck — APPROVED** [review]

**Pull request merged**